### PR TITLE
Retrieve comm body from s3 if passed in composed event

### PR DESCRIPTION
**Pull request by laurence-bird**

### CODE REVIEW

File: src/main/scala/com/ovoenergy/delivery/service/email/IssueEmail.scala [hunk 1]
```diff
@@ -1,44 +1,51 @@
...
-      } else {
-        Right(())
+  //TODO: Make single IssueComm, final tagless style algebra
+  def issue[F[_]](checkBlackWhiteList: (String) => BlackWhiteList.Verdict,
+                  isExpired: Option[Instant] => Boolean,
+                  content: CommContent[F],
+                  sendEmail: Content.Email => F[GatewayComm])(implicit F: Sync[F],
+                                                              clock: Clock): ComposedEmailV4 => F[GatewayComm] = {
+    composedEmail: ComposedEmailV4 =>
+      implicit val zdt = ZonedDateTime.ofInstant(clock.instant(), clock.getZone)
```
Comment (SystemFw) [inline src/main/scala/com/ovoenergy/delivery/service/email/IssueEmail.scala]: This is side-effectful (`clock.instant`)

File: src/main/scala/com/ovoenergy/delivery/service/persistence/DynamoPersistence.scala [hunk 2]
```diff
@@ -27,44 +27,43 @@ class DynamoPersistence(dbClient: AmazonDynamoDBAsync)(implicit config: DynamoDb
...
-            case Failure(error)           => cb(Right(Left(DynamoError(UnexpectedDeliveryError))))
+            case Success(Some(Right(_)))  => cb(Right(true))
+            case Success(None)            => cb(Right(false))
+            case Success(Some(Left(err))) => cb(Left(DynamoError(UnexpectedDeliveryError)))
+            case Failure(error)           => cb(Left(DynamoError(UnexpectedDeliveryError)))
           }
         }
       } catch {
         case e: AmazonDynamoDBException => {
           log.warn("Failed dynamoDb operation", e)
```
Comment (SystemFw) [inline src/main/scala/com/ovoenergy/delivery/service/persistence/DynamoPersistence.scala]: `delay`

File: src/test/scala/com/ovoenergy/delivery/service/email/mailgun/MailgunClientSpec.scala [hunk 3]
```diff
@@ -145,10 +156,19 @@ class MailgunClientSpec extends FlatSpec with Matchers with Arbitraries with Arb
...
-      result shouldBe Left(APIGatewayInternalServerError(EmailGatewayError))
-    }
+      MailgunClient
+        .sendEmail[IO](badResponse)
+        .apply(composedContent)
+        .attempt
+    }.toList
+
+    // TODO: Better way of doing this
+    f.sequence
```
Comment (SystemFw) [inline src/test/scala/com/ovoenergy/delivery/service/email/mailgun/MailgunClientSpec.scala]: There is no better way, sadly.

File: src/main/scala/com/ovoenergy/delivery/service/DeliverComm.scala [hunk 4]
```diff
@@ -7,22 +7,22 @@ import cats.implicits._
...
+    }
 
-      for {
-        isDuplicate <- dynamoPersistence.exists(commRecord)
-        gatewayComm <- Async[F].pure(isDuplicate.flatMap(issueCommIfUnique))
-        _           <- dynamoPersistence.persistHashedComm(commRecord)
-      } yield gatewayComm
+    for {
+      isDuplicate <- dynamoPersistence.exists(commRecord)
+      gatewayComm <- issueCommIfUnique(isDuplicate)
```
Comment (filosganga) [inline src/main/scala/com/ovoenergy/delivery/service/DeliverComm.scala]: I believe those two statements can be composed with `ifM`:
https://typelevel.org/cats/typeclasses/monad.html#ifm

I think something like:
```scala
dynamoPersistence.exists(commRecord).ifM(issueComm(event), F.unit)
```

But maybe I am saying a bs...:)

**Pull request merged**